### ISSUE TITLE
How to support Utterance alert of Array with two utteranceQueues?

### ISSUE THREAD

**Issue by zepumph** [phetsims/friction#204]
Utterance has a feature built into if for use in Friction in which you can supply a list of alerts, and it will cycle through them as it is alerted. This doesn't work well when you have code that tries to alert the same utterance in two different queues. Perhaps we should remove this list feature from Utterance

**Comment by zepumph** [comment]
I added a TODO to the exact spot in the code:

```js
  alertMaxTemp() {
    this.alert( () => {

      // TODO: Use of private function to alert to voicing and aria-live, see https://github.com/phetsims/friction/issues/204
      const alert = this.maxTempUtterance.getTextToAlert();
      phet.joist.sim.utteranceQueue.addToBack( alert );
      voicingUtteranceQueue.addToBack( alert );
    } );
  }
```

**Comment by zepumph** [comment]
@jessegreenberg and I discussed, and we like this direction. We agreed to keep the array of alerts in Utterance as a feature, and wish we had used it in RAP and JT recently. 

Don't forget to update the tests!

<details>

```diff
Index: friction/js/friction/view/describers/TemperatureIncreasingDescriber.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/friction/js/friction/view/describers/TemperatureIncreasingDescriber.js b/friction/js/friction/view/describers/TemperatureIncreasingDescriber.js
--- a/friction/js/friction/view/describers/TemperatureIncreasingDescriber.js	(revision 5059bf92e00a7b1f4600796145af689cb1264ddd)
+++ b/friction/js/friction/view/describers/TemperatureIncreasingDescriber.js	(date 1627066226188)
@@ -97,8 +97,7 @@
     this.tooSoonForNextAlert = false;
 
     this.maxTempUtterance = new Utterance( {
-      alert: [ MAX_TEMP_STRING, MAX_TEMP_STRING, resetSimMoreObservationSentenceString ],
-      loopAlerts: true
+      alert: [ MAX_TEMP_STRING, MAX_TEMP_STRING, resetSimMoreObservationSentenceString ]
     } );
 
     // @private
@@ -159,9 +158,13 @@
     this.alert( () => {
 
       // TODO: Use of private function to alert for voicing AND aria-live, see https://github.com/phetsims/friction/issues/204
-      const alert = this.maxTempUtterance.getTextToAlert();
+      const alert = this.maxTempUtterance.getTextToAlert( {
+        incrementCount: false,
+        loopAlerts: true
+      } );
       phet.joist.sim.utteranceQueue.addToBack( alert );
       voicingUtteranceQueue.addToBack( alert );
+      this.maxTempUtterance.incrementNumberOfTimesAlerted();
     } );
   }
 
Index: utterance-queue/js/Utterance.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/utterance-queue/js/Utterance.js b/utterance-queue/js/Utterance.js
--- a/utterance-queue/js/Utterance.js	(revision 4833b8a33018ceab2a8fd9ea69b011754af9eeb5)
+++ b/utterance-queue/js/Utterance.js	(date 1627066276903)
@@ -117,12 +117,17 @@
    * @returns {string}
    * @public (UtteranceQueue only)
    */
-  getTextToAlert() {
+  getTextToAlert( options ) {
+    options = merge( {
+      incrementCount: true,
+      loopAlerts: false
+    }, options );
     let alert;
     if ( typeof this._alert === 'string' ) {
       alert = this._alert;
     }
     else if ( this.loopAlerts ) {
+      assert && assert( Array.isArray( this._alert ) ); // sanity check
       alert = this._alert[ this.numberOfTimesAlerted % this._alert.length ];
     }
     else {
@@ -130,8 +135,15 @@
       const currentAlertIndex = Math.min( this.numberOfTimesAlerted, this._alert.length - 1 );
       alert = this._alert[ currentAlertIndex ];
     }
+    options.incrementCount && this.numberOfTimesAlerted++;
+    return alert;
+  }
+
+  /**
+   * @public
+   */
+  incrementNumberOfTimesAlerted() {
     this.numberOfTimesAlerted++;
-    return alert;
   }
 
   /**

```

**Comment by zepumph** [comment]
@jessegreenberg, the above patch was not really on point. I had forgotten during our meeting that `getTextToAlert` is actually private to sims, and so what was in place was just a work around. Thus it made more sense for our `incrementCount` to become an option to the Utterance, next to the current placement for loopAlerts. 


When I was adding tests for this new feature. I realized that this is very sketchy and buggy. The incrementing will only happen when the Utterance is plucked from the queue. That is an async step away from the operation, so most likely this isn't working as expected in Friction. I think we will need to discuss this further @jessegreenberg, because I'm feeling a bit stuck on this pattern. I am thinking that it is going to be a very important and popular thing to add a single Utterance to two different queues, so let's keep on this next week. Here is the patch.


<details>

```diff
Index: friction/js/friction/view/describers/TemperatureIncreasingDescriber.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/friction/js/friction/view/describers/TemperatureIncreasingDescriber.js b/friction/js/friction/view/describers/TemperatureIncreasingDescriber.js
--- a/friction/js/friction/view/describers/TemperatureIncreasingDescriber.js	(revision 82e7b43d6d332f4e2c36b6fdc9dc944f6cd8a656)
+++ b/friction/js/friction/view/describers/TemperatureIncreasingDescriber.js	(date 1627075503938)
@@ -98,7 +98,10 @@
 
     this.maxTempUtterance = new Utterance( {
       alert: [ MAX_TEMP_STRING, MAX_TEMP_STRING, resetSimMoreObservationSentenceString ],
-      loopAlerts: true
+      loopAlerts: true,
+
+      // since this alerts to description and voicing utteranceQueues at the same time, manually incremenet the count
+      incrementCountManually: true
     } );
 
     // @private
@@ -162,11 +165,11 @@
    */
   alertMaxTemp() {
     this.alert( () => {
+      phet.joist.sim.utteranceQueue.addToBack( this.maxTempUtterance );
+      voicingUtteranceQueue.addToBack( this.maxTempUtterance );
 
-      // TODO: Use of private function to alert for voicing AND aria-live, see https://github.com/phetsims/friction/issues/204
-      const alert = this.maxTempUtterance.getTextToAlert();
-      phet.joist.sim.utteranceQueue.addToBack( alert );
-      voicingUtteranceQueue.addToBack( alert );
+      // Increment once both have been alerted to.
+      this.maxTempUtterance.incrementNumberOfTimesAlerted();
     } );
   }
 
Index: utterance-queue/js/UtteranceTests.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/utterance-queue/js/UtteranceTests.js b/utterance-queue/js/UtteranceTests.js
--- a/utterance-queue/js/UtteranceTests.js	(revision 4833b8a33018ceab2a8fd9ea69b011754af9eeb5)
+++ b/utterance-queue/js/UtteranceTests.js	(date 1627075925452)
@@ -146,6 +146,63 @@
   testOrder( ', reset should start over' );
 } );
 
+QUnit.test( 'Utterance loopAlerts with incrementCountManually', async assert => {
+
+  const alert = new Utterance( {
+    alert: [ '1', '2', '3' ],
+    loopAlerts: true,
+    incrementCountManually: true,
+    alertStableDelay: 0 // we want to hear the utterance every time it is added to the queue
+  } );
+
+  const alert7 = async () => {
+    for ( let i = 0; i < 7; i++ ) {
+      utteranceQueue.addToBack( alert );
+      await timeout( sleepTiming );
+
+      // TODO: this fails if it is before the timeout!!! https://github.com/phetsims/friction/issues/204
+      alert.incrementNumberOfTimesAlerted();
+    }
+  };
+
+  const testOrder = messageSuffix => {
+
+    // newest at lowest index
+    assert.ok( alerts[ 6 ] === '1', `Array order1${messageSuffix}` );
+    assert.ok( alerts[ 5 ] === '2', `Array order2${messageSuffix}` );
+    assert.ok( alerts[ 4 ] === '3', `Array order3${messageSuffix}` );
+    assert.ok( alerts[ 3 ] === '1', `Array order4${messageSuffix}` );
+    assert.ok( alerts[ 2 ] === '2', `Array order5${messageSuffix}` );
+    assert.ok( alerts[ 1 ] === '3', `Array order6${messageSuffix}` );
+    assert.ok( alerts[ 0 ] === '1', `Array order7${messageSuffix}` );
+  };
+
+  await alert7();
+  testOrder( '' );
+  alert.reset();
+  await alert7();
+  testOrder( ', reset should start over' );
+
+
+  alert.reset();
+  alerts = [];
+  utteranceQueue.addToBack( alert );
+  await timeout( sleepTiming );
+  assert.ok( alerts[ 0 ] === '1', 'Array order 0' );
+  utteranceQueue.addToBack( alert );
+  await timeout( sleepTiming );
+  assert.ok( alerts[ 0 ] === '1', 'Array order 0 no increment' );
+  utteranceQueue.addToBack( alert );
+  await timeout( sleepTiming );
+  assert.ok( alerts[ 0 ] === '1', 'Array order 0 no increment again' );
+
+  alert.incrementNumberOfTimesAlerted();
+  utteranceQueue.addToBack( alert );
+  await timeout( sleepTiming );
+  assert.ok( alerts[ 0 ] === '2', 'Array order 1 after first increment' );
+
+} );
+
 QUnit.test( 'alertStable and alertStableDelay tests', async assert => {
   const highFrequencyUtterance = new Utterance( { alert: 'Rapidly Changing' } );
 
Index: utterance-queue/js/Utterance.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/utterance-queue/js/Utterance.js b/utterance-queue/js/Utterance.js
--- a/utterance-queue/js/Utterance.js	(revision 4833b8a33018ceab2a8fd9ea69b011754af9eeb5)
+++ b/utterance-queue/js/Utterance.js	(date 1627075366555)
@@ -51,6 +51,13 @@
       // to the beginning when complete. The default behavior (loopAlerts:false) is to repeat the last alert in the array until reset.
       loopAlerts: false,
 
+      // This option should only be used cautiously. Normally, it is a feature of Utterance to automatically count the
+      // times alerted, and to support cycling through alerts when `alert` is an array. This can be used in correlation
+      // with Utterance.incrementNumberOfTimesAlerted() to manually specify when it is time to increment. The use case
+      // for this option is when a single utterance wants to be provided to multiple utteranceQueues with the same alert.
+      // See https://github.com/phetsims/friction/issues/204 for original report.
+      incrementCountManually: false,
+
       // @returns {boolean} - if predicate returns false, the alert content associated
       // with this utterance will not be announced by a screen reader
       predicate: function() { return true; },
@@ -81,6 +88,7 @@
     // @private
     this._alert = options.alert;
     this.loopAlerts = options.loopAlerts;
+    this.incrementCountManually = options.incrementCountManually;
 
     // @public (read-only)
     this.numberOfTimesAlerted = 0; // keep track of the number of times alerted, this will dictate which alert to call.
@@ -123,6 +131,7 @@
       alert = this._alert;
     }
     else if ( this.loopAlerts ) {
+      assert && assert( Array.isArray( this._alert ) ); // sanity check
       alert = this._alert[ this.numberOfTimesAlerted % this._alert.length ];
     }
     else {
@@ -130,8 +139,15 @@
       const currentAlertIndex = Math.min( this.numberOfTimesAlerted, this._alert.length - 1 );
       alert = this._alert[ currentAlertIndex ];
     }
+    !this.incrementCountManually && this.numberOfTimesAlerted++;
+    return alert;
+  }
+
+  /**
+   * @public
+   */
+  incrementNumberOfTimesAlerted() {
     this.numberOfTimesAlerted++;
-    return alert;
   }
 
   /**

**Comment by zepumph** [comment]
For now, I would prefer a better work around, I'll just have 2 Utterance instances, one per queue. This will help with `cancelOther` features and alertStableDelay etc. Especially for https://github.com/phetsims/friction/issues/207

**Comment by zepumph** [comment]
After another discussion with @jessegreenberg, we realized there are many things about Utterance that are buggy when you add an Utterance to multiple queues, see https://github.com/phetsims/utterance-queue/issues/20. On hold until we handle things over there.

**Comment by zepumph** [comment]
Now that we have no usages of arrays as utterance alerts in this code, we can close this issue. There is just some cleanup to do in common code over in https://github.com/phetsims/utterance-queue/issues/20. Closing
